text/0048-schema-annotations.md

```diff
@@ -0,0 +1,140 @@
+Thus the full schema syntax becomes:
+```
+Schema      := {Namespace}
+Namespace   := ('namespace' Path '{' {Decl} '}') | {Decl}
+Decl        := Entity | Action | TypeDecl
+Entity      := Annotations 'entity' Idents ['in' EntOrTyps] [['='] RecType] ';'
+Action      := Annotations 'action' Names ['in' (Name | '[' [Names] ']')] [AppliesTo] [ActAttrs]';'
+TypeDecl    := Annotations 'type' IDENT '=' Type ';'
+Type        := PRIMTYPE | IDENT | SetType | RecType
+EntType     := Path
+SetType     := 'Set' '<' Type '>'
+RecType     := '{' [AttrDecls] '}'
+AttrDecls   := Name ['?'] ':' Type [',' | ',' AttrDecls]
+AppliesTo   := 'appliesTo' '{' AppDecls '}'
+ActAttrs    := 'attributes' '{' AttrDecls '}'
+AppDecls    := ('principal' | 'resource') ':' EntOrTyps [',' | ',' AppDecls]
+             | 'context' ':' RecType [',' | ',' AppDecls]
+Path        := IDENT {'::' IDENT}
+EntTypes    := Path {',' Path}
+EntOrTyps   := EntType | '[' [EntTypes] ']'
+Name        := IDENT | STR
+Names       := Name {',' Name}
+Idents      := IDENT {',' IDENT}
+Annotation  := '@' IDENT '(' STR ')
+Annotations := Annotation {Annotations}
+
+IDENT       := ['_''a'-'z''A'-'Z']['_''a'-'z''A'-'Z''0'-'9']* - PRIMTYPE
+STR         := Fully-escaped Unicode surrounded by '"'s
+PRIMTYPE    := 'Long' | 'String' | 'Bool'
+WHITESPC    := Unicode whitespace
+COMMENT     := '//' ~NEWLINE* NEWLINE
+```
+
+### JSON Syntax
+None of the three top-level constructs (EntityTypes, Actions, CommonTypes) in schemas allow for arbitrary key/value pairs. 
+This means a new key can be safely added while preserving backwards compatibility.
+This proposal reserves the `annotations` key at the top level of each of those constructs, which contains an Object, containing each annotation key as an Object key, associated with the annotation value.
+The only oddness here is Common Types, whose toplevel is a regular type. While this should still be backwards compatible, it will look a little odd to have annotations in some types and not in others.
+
+### Human Readable Syntax
+
+## Drawbacks
+
+1. Complexity: adds more complexity to schema
+2. Oddness around syntax for Common Types in JSON form
+3. By not taking a stance on annotation meanings, it makes it harder for a standard to form around them (ex: for doc strings)
+4. Multi-line docstrings are technically valid but awkward.
+
+## Alternatives
+
+### Take a stance
+Reverse our decision around annotations and start taking stances on what annotations mean.
+This lets us standardize certain annotations, like `doc`. 
+This probably can't happen unless we also do this for policies, which we've said we don't want to do.
+### Doc Strings as comments
+Instead of annotations, we could add "doc-strings" as a first class feature.
+Could look like this:
+```
+/# Stop users from accessing a high security document unless:
+/#  A) The principal and user are at the same location
+/#  B) The principal has a job level greater than  4
+forbid(principal, action, resource) when {
```

Comment:
No, the docs case was what was asked for by someone in the slack, and I was seeing if people preferred the more general approach

Reply:
With the customer schema syntax I do think we'll see more requests for annotations in the schema eventually. It would feel natural to use the same annotation mechanism in policies and schema.

Reply:
No reason we can't do both annotations and doc comments

Reply:
In Rust, doc comments are actually just sugar for Rust's equivalent of annotations.  (Specifically, sugar for the [`#[doc]` attribute](https://doc.rust-lang.org/rustdoc/write-documentation/the-doc-attribute.html).)  We could consider doing something similar.  What that means for this RFC is that we could support just doc comments now, and later if we introduce general annotations we could have doc comments desugar into a special reserved annotation.  Or, we could support general annotations now, and later we could introduce doc comments as a sugar on top.  We don't have to worry about closing off one or the other path.

Reply:
For a doc comment syntax, I think `//#` is preferable to `/#` since backwards compatible with valid line comments, and similar to other syntaxes for overloading comments.   An example is `//!` , the esbuild [legal comment](https://esbuild.github.io/api/#legal-comments) syntax to preserve in output files .

```
//! Copyright Cedar Contributors
//! SPDX-License-Identifier: Apache-2.0
```